Gems/Terrain/Code/Source/TerrainRenderer/TerrainClipmapManager.h

```diff
@@ -22,35 +24,69 @@ namespace Terrain
         AZ_RTTI(TerrainClipmapManager, "{5892AEE6-F3FA-4DFC-BBEC-77E1B91506A2}");
         AZ_DISABLE_COPY_MOVE(TerrainClipmapManager);
 
-        TerrainClipmapManager() = default;
+        TerrainClipmapManager();
         virtual ~TerrainClipmapManager() = default;
 
+        enum ClipmapName : uint32_t
+        {
+            MacroColor = 0,
+            MacroNormal,
+            DetailColor,
+            DetailNormal,
+            DetailHeight,
+            // Miscellany clipmap combining:
+            // roughness, specularF0, metalness, occlusion
+            DetailMisc,
+
+            Count
+        };
+
+        // Shader input names
+        static constexpr const char* ClipmapDataShaderInput = "m_clipmapData";
+        static constexpr const char* ClipmapImageShaderInput[ClipmapName::Count] = {
+            "m_macroColorClipmaps",
+            "m_macroNormalClipmaps",
+            "m_detailColorClipmaps",
+            "m_detailNormalClipmaps",
+            "m_detailHeightClipmaps",
+            "m_detailMiscClipmaps"
+        };
+
         void Initialize(AZ::Data::Instance<AZ::RPI::ShaderResourceGroup>& terrainSrg);
         bool IsInitialized() const;
         void Reset();
         bool UpdateSrgIndices(AZ::Data::Instance<AZ::RPI::ShaderResourceGroup>& srg);
 
         void Update(const AZ::Vector3& cameraPosition, AZ::Data::Instance<AZ::RPI::ShaderResourceGroup>& terrainSrg);
 
-        AZ::Data::Instance<AZ::RPI::AttachmentImage> GetClipmapImage(const AZ::Name& clipmapName) const;
+        // Import the clipmap to the frame graph and set scope attachment access,
+        // so that the compute pass can build dependencies accordingly.
+        void ImportClipmap(ClipmapName clipmapName, AZ::RHI::FrameGraphAttachmentInterface attachmentDatabase) const;
+        void UseClipmap(ClipmapName clipmapName, AZ::RHI::ScopeAttachmentAccess access, AZ::RHI::FrameGraphInterface frameGraph) const;
+
+        AZ::Data::Instance<AZ::RPI::AttachmentImage> GetClipmapImage(ClipmapName clipmapName) const;
     private:
         void UpdateClipmapData(const AZ::Vector3& cameraPosition);
         void InitializeClipmapData();
         void InitializeClipmapImages();
 
-        static constexpr uint32_t ClipmapStackSize = 5;
+        static constexpr uint32_t MacroClipmapStackSize = 3;
+        static constexpr uint32_t DetailClipmapStackSize = 5;
         static constexpr uint32_t ClipmapSizeWidth = 1024;
         static constexpr uint32_t ClipmapSizeHeight = 1024;
 
+        static_assert(DetailClipmapStackSize > MacroClipmapStackSize,
+            "To avoid seams, macro clipmaps will use a lower resolution, so that it can take advantage of free seam blending.");
+
```

Comment:
Is this really required for Detail stack size to be > Macro? Is there any harm in removing this restriction? IMO it seems more like a recommendation than a requirement.

Reply:
Because I am using the detail stack size to construct arrays, it will crash if macro size is bigger. And it doesn't really make sense if macro material has higher resolutions than detail material.